runtime/src/main/java/org/corfudb/runtime/view/replication/IReplicationProtocol.java

```diff
@@ -63,49 +69,62 @@
-     * @param globalAddresses       A list of addresses to read from.
-     * @return                      A map of addresses to committed
-     *                              addresses, hole filling if necessary.
+     * @param runtimeLayout the RuntimeLayout stamped with layout to use for the read.
+     * @param addresses     a list of addresses to read from.
+     * @param waitForWrite  flag whether wait for write is required or hole fill directly.
+     * @return a map of addresses to data commit at these address, hole filling if necessary.
      */
     default @Nonnull
-            Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout, List<Long> globalAddresses) {
-        return globalAddresses.parallelStream()
-                .map(a -> new AbstractMap.SimpleImmutableEntry<>(a, read(runtimeLayout, a)))
-                .collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue()));
+    Map<Long, ILogData> readAll(RuntimeLayout runtimeLayout,
+                                List<Long> addresses,
+                                boolean waitForWrite) {
+        return addresses.parallelStream()
+                .map(addr -> new SimpleImmutableEntry<>(addr, read(runtimeLayout, addr)))
+                .collect(Collectors.toMap(SimpleImmutableEntry::getKey, SimpleImmutableEntry::getValue));
     }
 
-    /** Read data from a range.
+    /**
+     * Read data in the given range.
      *
-     * <p>This method functions exactly like a readAll, except
-     * that it returns the result for a range of addresses.
+     * <p>This method functions exactly like a read, except
+     * that it returns the result for multiple addresses.
      *
      * <p>An implementation may optimize for this type of
      * bulk request, but the default implementation
      * just performs multiple reads (possible in parallel).
      *
-     * @param runtimeLayout         The RuntimeLayout stamped with layout to use for the readRange.
-     * @param globalAddresses       A list of addresses to read from.
-     * @return                      A map of addresses to committed
-     *                              addresses, hole filling if necessary.
+     * @param runtimeLayout the RuntimeLayout stamped with layout to use for the read.
+     * @param addresses     a range of addresses to read from.
+     * @param waitForWrite  flag whether wait for write is required or hole fill directly.
+     * @return a map of addresses to data commit at these address, hole filling if necessary.
      */
-    Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout, Set<Long> globalAddresses);
-    /** Peek data from a given address.
+    default @Nonnull
+    Map<Long, ILogData> readRange(RuntimeLayout runtimeLayout,
+                                Range<Long> addresses,
+                                boolean waitForWrite) {
+        return ContiguousSet.create(addresses, DiscreteDomain.longs())
+                .parallelStream()
```

Comment:
Remove this default implementation. This usage of parallelStream can be problematic.

Reply:
Done.